pkg/api/customization/vsphere/handler.go

```diff
@@ -0,0 +1,187 @@
+}
+
+func NewVsphereHandler(scaledContext *config.ScaledContext) http.Handler {
+	return &handler{
+		schemas:       scaledContext.Schemas,
+		secretsLister: scaledContext.Core.Secrets(namespace.GlobalNamespace).Controller().Lister(),
+	}
+}
+
+func (v *handler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	res.Header().Set("Content-Type", "application/json")
+	fieldName := mux.Vars(req)["field"]
+	dc := req.FormValue("dataCenter")
+
+	if fieldName == "" || !validFieldName(fieldName) {
+		util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: invalid field name", httperror.NotFound.Code))
+		return
+	}
+
+	cc, err := v.getCloudCredential(req)
+	if err != nil {
+		util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: %s", httperror.InvalidBodyContent.Code, err.Error()))
+		return
+	}
+
+	var data []string
+	if fieldNames[fieldName] == SoapFinder {
+		data, err = processSoapFinder(req.Context(), fieldName, cc, dc)
+		if err != nil {
+			util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: %s", httperror.InvalidBodyContent.Code, err.Error()))
+			return
+		}
+	}
+
+	if fieldNames[fieldName] == ContentLibraryManager {
+		l := req.FormValue("library")
+		data, err = processContentLibraryManager(req.Context(), fieldName, cc, l)
+		if err != nil {
+			util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: %s", httperror.InvalidBodyContent.Code, err.Error()))
+			return
+		}
+	}
+
+	if fieldNames[fieldName] == TagsManager {
+		c := req.FormValue("category")
+		data, err = processTagsManager(req.Context(), fieldName, cc, c)
+		if err != nil {
+			util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: %s", httperror.InvalidBodyContent.Code, err.Error()))
+			return
+		}
+	}
+
+	if fieldNames[fieldName] == CustomFieldsFinder {
+		data, err = processCustomFieldsFinder(req.Context(), cc)
+		if err != nil {
+			util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: %s", httperror.InvalidBodyContent.Code, err.Error()))
+			return
+		}
+	}
+
+	js, err := json.Marshal(map[string][]string{
+		"data": data,
+	})
+
+	if err != nil {
+		util.ReturnHTTPError(res, req, httperror.NotFound.Status, fmt.Sprintf("%s: %s", httperror.ServerError.Code, err.Error()))
+		return
+	}
+
+	res.WriteHeader(http.StatusOK)
+	res.Write(js)
+}
+
+func (v *handler) getCloudCredential(req *http.Request) (*corev1.Secret, error) {
+	credID := req.FormValue("cloudCredentialId")
+	if credID == "" {
+		return nil, fmt.Errorf("cloud credential required")
+	}
+
+	var accessCred client.CloudCredential
+	if err := access.ByID(v.generateAPIContext(req), &schema.Version, client.CloudCredentialType, credID, &accessCred); err != nil {
+		if apiError, ok := err.(*httperror.APIError); ok {
+			if apiError.Code.Status == httperror.PermissionDenied.Status || apiError.Code.Status == httperror.NotFound.Status {
+				return nil, fmt.Errorf("cloud credential not found")
```

Comment:
The errors in here need to be the same basically. Otherwise this will leak if a cloud cred exists by that ID or not. If you can't find it or they can't access it, the error should be the same. If there is a legit error that can be returned.